argparse

monyarm opened the pull request:
- **feat(mcl): Add commandline argument support**
- **feat(mcl/utils/log): Implement errorAndExit function**
- **feat(mcl/utils/process): Add support to execute function, for supplying redirect; Improve execute function argument escaping, to not escape globs**
- **feat(mcl/config): Implement mcl config command**
- **refactor(mcl/dub): Remove deprecated dflags for improved build configuration**
- **fix(ci_matrix): Initialize params directly in ci_matrix function**
- **fix(utils/cachix): Modify getCachixDeploymenApiUrl definition, to allow compilation with newer d compiler versions.**
- **feat(mcl/main): Integrate argparse for command-line argument parsing and enhance command handling**
- **feat(commands/get_fstab): Enhance get_fstab command with argparse integration and structured argument handling**
- **feat(commands/deploy_spec): Enhance deploy_spec command with argparse integration and structured argument handling**
- **feat(commands/host_info): Enhance host_info command with argparse integration and structured argument handling**
- **feat(commands/config): Enhance config command with argparse integration and structured argument handling**
- **feat(commands/machine): Enhance machine command with argparse integration and structured argument handling**
- **feat(commands): Enhance ci/matrix commands with argparse integration and structured argument handling where applicable**

monyarm committed:
refactor(mcl/dub): Remove deprecated dflags for improved build configuration

monyarm committed:
fix(ci_matrix): Initialize params directly in ci_matrix function

monyarm committed:
feat(mcl/main): Integrate argparse for command-line argument parsing and enhance command handling

monyarm committed:
feat(commands/get_fstab): Enhance get_fstab command with argparse integration and structured argument handling

monyarm committed:
feat(commands/deploy_spec): Enhance deploy_spec command with argparse integration and structured argument handling

monyarm committed:
feat(commands/host_info): Enhance host_info command with argparse integration and structured argument handling

monyarm committed:
feat(commands/config): Enhance config command with argparse integration and structured argument handling

monyarm committed:
feat(commands/machine): Enhance machine command with argparse integration and structured argument handling

monyarm committed:
feat(commands): Enhance ci/matrix commands with argparse integration and structured argument handling where applicable

monyarm committed:
fix(actions): Correct command names in GitHub workflows and action configuration

monyarm committed:
fix(commands/machine): Correct command name in createMachine function for host info retrieval

monyarm committed:
feat(mcl): Add command aliases for backwards compatibility

reo101 commented on packages/mcl/src/main.d:
```diff
@@ -1,73 +1,89 @@
+            "print_table_args,"~
+            "shard_matrix_args,"~
+            "Default!unknown_command_args"~
+        ") cmd;";
```
Could this be done in a similar fashion to the following `genSubCommandMatch` template?

PetarKirov replied:
This was refactored in 931cd5abea4b542248b72bf69364b23f74930823.

reo101 commented on packages/mcl/src/main.d:
```diff
@@ -1,73 +1,89 @@
-    LogLevel logLevel = LogLevel.info;
-    args.getopt("log-level", &logLevel);
+@(Command(" ").Description(" "))
+struct unknown_command_args {}
```
Prefer capitalized names ([`PascalCase`](https://dlang.org/dstyle.html#naming_classes), to be exact) for `struct`s. Applies to the rest of the `_args` `struct`s too.

PetarKirov replied:
Fixed in 0f9a636a2fa0e8b44ab77ba7d53dffff7975eca6.

reo101 commented on packages/mcl/src/src/mcl/commands/ci_matrix.d:
```diff
@@ -186,13 +187,20 @@ Package[] checkCacheStatus(Package[] packages)
+{
+}
+
+export int print_table(print_table_args args)
```
Can we instead try supporting argument-less commands by providing an aptly argument-less function returning `void`, with `genSubCommandArgs` being updated to dynamically check whether or not each sub-command handler function has arguments/return value at all before taking them into account?
This also applies to all `_args` `struct`s annotated with `@(Command(" ").Description(" "))`.

PetarKirov replied:
I don't think so, as as [argparse `SubCommand`s](https://andrey-zherikov.github.io/argparse/subcommands.html) are expressed as `SumType`s.

We could actually do something like this:

```d
struct Command1
{
    // no CLI args
    
    int run() { /* ... */ }
}

struct Command2
{
    int arg1;
    string arg2;
    
    int run()
    {
        // Use `this.arg1` and `this.arg2`... 
    }
}
```

PetarKirov commented on packages/mcl/src/src/mcl/commands/config.d:
```diff
@@ -14,112 +17,166 @@ import mcl.utils.nix : nix, queryStorePath;
+        home_apply_args,
+        home_edit_args,
+        Default!unknown_command_args
+    ) cmd;
+}
+
+@(Command("apply").Description("Apply user configuration"))
+struct home_apply_args
+{
+    @(PositionalArgument(0).Placeholder("desktop/server").Description("Type of home configuration"))
+    string type;
+}
+
+@(Command("edit").Description("Edit user configuration"))
+struct home_edit_args
+{
+    @(PositionalArgument(0).Placeholder("desktop/server").Description("Type of home configuration"))
+    string type;
+}
```
```suggestion
enum HomeConfigType { desktop, server }

@(Command("apply").Description("Apply user configuration"))
struct home_apply_args
{
    @(PositionalArgument(0).Placeholder("desktop/server").Description("Type of home configuration"))
    HomeConfigType type;
}

@(Command("edit").Description("Edit user configuration"))
struct home_edit_args
{
    @(PositionalArgument(0).Placeholder("desktop/server").Description("Type of home configuration"))
    HomeConfigType type;
}
```

PetarKirov commented on packages/mcl/src/src/mcl/commands/config.d:
```diff
@@ -14,112 +17,166 @@ import mcl.utils.nix : nix, queryStorePath;
 }
 
-void sys(string[] args)
+int apply(string type, string value)
```
I suggest having two separate functions:

```d
int switchSystem(string machineName);
int switchHome(HomeConfigType type, string username);
```

The comment about `switchHome` is dependent on this change in the Justfile:

<img width="438" height="97" alt="Image" src="https://github.com/user-attachments/assets/997de006-d069-4452-8d20-c9b523fa5228" />

PetarKirov commented on packages/mcl/src/src/mcl/commands/deploy_spec.d:
```diff
@@ -51,9 +58,11 @@ export void deploy_spec(string[] args)
 
     if (!spec.agents.length)
-        return;
+        return 1;
```
Having no configurations to deploy is ok:

```suggestion
        return 0;
```

PetarKirov commented on packages/mcl/src/src/mcl/commands/host_info.d:
```diff
@@ -72,13 +75,15 @@ export void host_info(string[] args)
     coda.uploadHostInfo(hostInfo);
 
+    return 1;
+
```
```suggestion
    return 0;
```

PetarKirov commented on packages/mcl/src/src/mcl/commands/host_info.d:
```diff
@@ -59,7 +59,10 @@ string[string] getProcInfo(string fileOrData, bool file = true)
 
-export void host_info(string[] args)
+@(Command("host-info", "host_info").Description("Get information about the host machine"))
+struct host_info_args {}
```
```suggestion
struct HostInfoArgs
{
    bool uploadToCoda;
}
```

PetarKirov commented on packages/mcl/src/src/mcl/commands/host_info.d:
```diff
@@ -72,13 +75,15 @@ export void host_info(string[] args)
 
     if (!params.codaApiToken) {
```
```suggestion
    if (args.uploadToCoda && !params.codaApiToken) {
```

PetarKirov commented on packages/mcl/src/src/mcl/commands/machine.d:
```diff
@@ -365,43 +368,58 @@ struct HardwareConfiguration {
+    @(NamedArgument(["description"]).Placeholder("description").Description("Description of the user"))
+    string description;
+    @(NamedArgument(["is-normal-user"]).Placeholder("true/false").Description("Is this a normal user"))
+    bool isNormalUser;
```
All users are non-system, so we don't need this option.

PetarKirov committed:
fix(mcl.commands): Fix function signature of `ci` and `ci_matrix` commands

PetarKirov committed:
refactor(mcl.commands): Improve meta-programming to automatically define top-level commands

PetarKirov committed:
fix(mcl.commands.config): Update nixos config repo names

reo101 committed:
feat(mcl): `PascalCase` all command arg structs

PetarKirov committed:
refactor(mcl.commands): Use selective import for `argparse`

PetarKirov committed:
style(mcl.commands): Reformat command definition

PetarKirov committed:
refactor(mcl.env): Remove the requirement of a `setup` function

PetarKirov committed:
fix(mcl.commands.shard_matrix): Fallback to `GITHUB_OUTPUT` env variable

PetarKirov committed:
feat(mcl): Upgrade `argparse`: `1.4.1` -> `2.0.2`

* Switch to nixpkgs version of `buildDubPackage`
* Use `nix-to-dub` to generate dub lock file:

    nix run nixpkgs#dub-to-nix > dub-lock.json

* Enable `checkPhase` (default is off in nixpkgs, unlike dlang.nix)
* Escape flags disabled tests flag passed to `dubTestFlags`
* Add installPhase

PetarKirov committed:
Revert "fix(ci_matrix): Initialize params directly in ci_matrix function"

This reverts commit 0fdda48709620fbe2dd28776cb0d5a8090355b5e.

monyarm committed:
refactor(): Remove Params and switch it to argparse

monyarm committed:
fix(): Review Compliance

PetarKirov committed:
fix(mcl): Workaround breaking change in handling of `dubTestFlags` in `buildDubPackage`

PR https://github.com/NixOS/nixpkgs/pull/344744 introduced a breaking change
in the way `dubTestFlags` are passed to `dub-check-hook.sh`.

PetarKirov committed:
build(mcl): Skip stripping debug info for better stacktraces

PetarKirov committed:
feat(mcL): Improve `mcl ci` impl - build & push in parallel
